var/spack/repos/builtin/packages/silo/package.py

```diff
@@ -55,7 +54,9 @@ class Silo(AutotoolsPackage):
 
     patch('remove-mpiposix.patch', when='@4.8:4.10.2')
     patch('H5FD_class_t-terminate.patch', when='@:4.10.2 ^hdf5@1.10.0:')
+    patch('H5EPR_SEMI_COLON.patch', when='@:4.11 ^hdf5@1.10.8:1.11,1.12.1:1.13')
```

Comment:
As written, this appears carefully constructed to avoid `@1.12.0`, but I don't see such a version in the `hdf5` package, so if that is OK then this could be simplified to:

`when='@:4.11 ^hdf5@1.10.8:1.13')`

I could be missing something though: this seems related to https://github.com/spack/spack/pull/27613#issuecomment-990364806 but I should say that the `when` statement as of now doesn't specify either of the ranges described there:

> Silo 4.10.2 (default) builds with HDF5 versions 1.8 and 1.10 <= 1.10.7 needs patches for 1.10.8 – 1.10.x, 1.12, 1.13
> Silo 4.11 builds with HDF5 versions 1.8 and 1.10 <= 1.10.7, 1.12.0, needs a (subset) patch for 1.10.8 – 1.10.x and 1.12.1-1.12.x, and augmented (subset) patch for 1.13

For example, `hdf5@1.10.8:1.11` means "any version between 1.10.8 and 1.11.x for any x, inclusively" (so effectively, any version before 1.12.0). If you want to apply one of the bounds exactly as described, I can translate that into a version range for you. IMO it might make sense to constrain what you are willing to build though and create a simpler `when` statement (in other words, you can constrain the when to match what you are building with, and you don't need to run multiple builds to find the exact conditions where the patch needs to be applied).

Reply:
> As written, this appears carefully constructed to avoid `@1.12.0`, but I don't see such a version in the `hdf5` package, so if that is OK then this could be simplified to:
> 

line 40 of hdf5/package.py:
      version('1.12.0', sha256='a62dcb276658cb78e6795dd29bf926ed7a9bc4edf6e77025cd2c689a8f97c17a')

> `when='@:4.11 ^hdf5@1.10.8:1.13')`
> 
> I could be missing something though: this seems related to [#27613 (comment)](https://github.com/spack/spack/pull/27613#issuecomment-990364806) but I should say that the `when` statement as of now doesn't specify either of the ranges described there:
> 
> > Silo 4.10.2 (default) builds with HDF5 versions 1.8 and 1.10 <= 1.10.7 needs patches for 1.10.8 – 1.10.x, 1.12, 1.13
> > Silo 4.11 builds with HDF5 versions 1.8 and 1.10 <= 1.10.7, 1.12.0, needs a (subset) patch for 1.10.8 – 1.10.x and 1.12.1-1.12.x, and augmented (subset) patch for 1.13
> 
> For example, `hdf5@1.10.8:1.11` means "any version between 1.10.8 and 1.11.x for any x, inclusively" (so effectively, any version before 1.12.0). If you want to apply one of the bounds exactly as described, I can translate that into a version range for you. IMO it might make sense to constrain what you are willing to build though and create a simpler `when` statement (in other words, you can constrain the when to match what you are building with, and you don't need to run multiple builds to find the exact conditions where the patch needs to be applied).

I'm not following all of this, but the silo patch is needed for silo@4.11 with current hdf5 versions 1.10.8 and 1.12.1.  The patch is not needed for silo@4.11 with hdf5 1.8, hdf51.10.0 - 1.10.7, or hdf5 1.12.0.  The hdf5 change that necessitates the H5EPR_SEMI_COLON.patch in silo was introduced in versions 1.10.8 and 1.12.1 after version 1.12.0 had been released.  

The statement in silo/package.py "conflicts('hdf5@1.10.8:', when="@:4.10.2") results in spack choosing silo@4.11 instead of 4.10.2 (the default version) for hdf5 versions 1.10.8 and above

Prior to the experimental 1.13.0 release, there were no 1.x.y releases of hdf5 where x was an odd number.  There will not be any 1.11.x releases. The patch will be needed for any future 1.10.x or 1.12.x releases, which are both planned.  hdf5 1.13.0 has changes that will require either a different patch or a new silo release version, which Mark Miller said will wait until after the holidays.

I will happily consider and try any translated or simplified when statement that you propose.  The H5EPR_SEMI_COLON.patch is needed for silo@4.11 hdf5@1.10.x where x >=8 and hdf51.12.x where x >= 1.

Reply:
> The patch will be needed for any future 1.10.x or 1.12.x releases, which are both planned. hdf5 1.13.0 has changes that will require either a different patch or a new silo release 

Thanks for the explanation! In that case I think the statement will need one minor change, from 

```
when='@:4.11 ^hdf5@1.10.8:1.11,1.12.1:1.13'
```

to

```
when='@:4.11 ^hdf5@1.10.8:1.11,1.12.1:1.12'
```

Only the last character changed, but FYI upper bounds are not exclusive, and in our syntax for a version range, `x.y` matches any `x.y.z` (so currently this applies for all hdf5 1.13.z versions). You might find it more straightforward to write:

```
when='@:4.11 ^hdf5@1.10.8:1.11,1.12.1:1.12.999'
```

Reply:
Thank you! @scheibelp 
I will update the statement with a comment regarding the intent as follows:
    # H5EPR_SEMI_COLON.patch should be applied only to silo@4.11 when building
    # with hdf5@1.10.8 or later 1.10 or with hdf5@1.12.1 or later 1.12
    patch('H5EPR_SEMI_COLON.patch', when='@:4.11 ^hdf5@1.10.8:1.10,1.12.1:1.12')
I'll go with 1.10 and 1.12 for the upper bounds to have an example of how this works for future reference (there are no 1.11.x releases).